improver/lapse_rate.py

```diff
@@ -396,75 +466,38 @@ def process(self, temperature_cube, orography_cube, land_sea_mask_cube):
-
-            temperature_data = temp_slice.data
-
-            # Fill sea points with NaN values. Can't use Numpy mask since not
-            # recognised by "generic_filter" function.
-            temperature_data = np.where(land_sea_mask, temperature_data,
-                                        np.nan)
-
-            # Saves all neighbourhoods into "all_temp_subsections".
-            # cval is value given to points outside the array.
-            fnc = SaveNeighbourhood(allbuffers=all_temp_subsections)
-            generic_filter(temperature_data, fnc.filter,
-                           size=self.nbhood_size,
-                           mode='constant', cval=np.nan)
-
-            fnc = SaveNeighbourhood(allbuffers=all_orog_subsections)
-            generic_filter(orography_data, fnc.filter,
-                           size=self.nbhood_size,
-                           mode='constant', cval=np.nan)
-
-            # height_diff_mask is True for points where the height
-            # difference between the central point and its neighbours
-            # is > max_height_diff.
-            height_diff_mask = self._create_heightdiff_mask(
-                all_orog_subsections)
-
-            # Mask points with extreme height differences as NaN.
-            all_orog_subsections = np.where(height_diff_mask, np.nan,
-                                            all_orog_subsections)
-            all_temp_subsections = np.where(height_diff_mask, np.nan,
-                                            all_temp_subsections)
-
-            # Loop through both arrays and find gradient of each subsection.
-            # The gradient indicates lapse rate - save into another array.
-            # TODO: This for loop is the bottleneck in the code and needs to
-            # be parallelised.
-            lapse_rate_array = [self._calc_lapse_rate(temp, orog)
-                                for temp, orog in zip(all_temp_subsections,
-                                                      all_orog_subsections)]
-
-            lapse_rate_array = np.array(
-                lapse_rate_array, dtype=np.float32).reshape(dataarray_shape)
-
-            # Enforces upper and lower limits on lapse rate values.
-            lapse_rate_array = np.where(lapse_rate_array < self.min_lapse_rate,
-                                        self.min_lapse_rate, lapse_rate_array)
-            lapse_rate_array = np.where(lapse_rate_array > self.max_lapse_rate,
-                                        self.max_lapse_rate, lapse_rate_array)
-
-            lapse_rate_slice.data = lapse_rate_array
-            lapse_rate_cube_list.append(lapse_rate_slice)
-
-        lapse_rate_cube = lapse_rate_cube_list.merge_cube()
-        lapse_rate_cube.rename('air_temperature_lapse_rate')
-        lapse_rate_cube.units = 'K m-1'
+        # Create list of arrays over "realization" coordinate
+        has_realization_dimension = False
+        if temperature_cube.coords("realization"):
```

Comment:
Not quite.  It doesn't want to put the data in an array if there is a scalar realization dimension.  That's why this loop needs to catch whenever there is a realization coordinate, even if it is not dimensional.

This looping is complicated and horrible because we expect even a single realization to be a dimension coordinate. The whole code base is written assuming that.  But our inputs might have a scalar realization as scalar or dimensional.  I wanted to address this problem in a follow-on PR as there are a few places this has had to be overly complex.  Your solution may well be correct but I'd really prefer to test it in isolation from all the other changes.

Reply:
Ah turns out you are right.  Now have unit test to cover this properly.